doc/developer/design/20230315_backup_restore.md

```diff
@@ -0,0 +1,225 @@
+
+While this imposes some additional cost, it’s modest both in absolute terms and relative to our other options. See [the appendix](#appendix-a-s3-costs) for more details.
+
+## Whole-environment backups
+
+### Correctness
+
+Unlike for a single shard, the state of an entire environment can not be represented with a single number. (Including the `mztimestamp` — the timestamp of a shard can lag arbitrarily behind, if eg. the cluster that populates that shard is crashing.) Conceptually, an environment backup is a mapping between shard id and the seqno of the per-shard backup. (Along with any additional environment-level state, like the stash.)
+
+Correctness criteria are also a bit more subtle, and involve reasoning about the dependencies between collections. We’re aware of two important correctness properties:
+
+1. **No broken since holds.** If B depends on A, B will maintain some hold on the **since** of A. (Commonly, this will mean that A’s since can’t advance past B’s upper.)
+2. **No time travel.** If B depends on A, B’s upper should not be past A’s upper. (If A is a table and gets restored to some old state, B shouldn’t contain any of the updates from A that no longer exist.)
+
+In a running environment, property 1 is carefully enforced by the various controllers, and property 2 is a consequence of causality. We need to ensure make sure that our restore process respects these properties as well.
+
+### Implementation
+
+Our proposed restore implementation relies on the ordering properties guaranteed by CRDB:
+
+- Restore the entire CRDB backup for an environment, including the Persist-shard metadata and the catalog/stash.
+- For every shard in the environment, run the per-shard restore process described above.
+
+This is straightforward to implement, and does not require any significant changes to the running Materialize environment.
+
+This is not _strictly_ safe, since CRDB is not a strictly serializable store; in particular, the docs [describe the risk of a causal reverse](https://www.cockroachlabs.com/blog/consistency-model/#as-of-system-timequeries-and-backups), where a read concurrent with updates A and B may observe B but not A, even if B is a real-time consequence of A. In the context of Materialize, if A is a compare-and-append and B is a compare-and-append on some downstream shard, a causal reverse could cause our backup of B to contain newer data than our backup of A, breaking correctness requirement 2.
+
+A causal reverse is expected to be very rare, and if encountered we can just try restoring another backup. However, [see below for a discussion of the alternatives](#conclusion-and-alternatives).
+
+# Rollout
+
+We already capture CRDB changefeeds in some environments; we’ll need to enable this everywhere, and to make sure we’re choosing our retention carefully.
```

Comment:
Seems like this is out of date? (Talks about changefeeds instead of CRDB backups)